velox/expression/tests/ArrayProxyTest.cpp

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glog/logging.h"
+#include "gtest/gtest.h"
+#include "velox/expression/VectorUdfTypeSystem.h"
+#include "velox/functions/Udf.h"
+#include "velox/functions/prestosql/tests/FunctionBaseTest.h"
+
+namespace facebook::velox {
+namespace {
+// Function that creates array with values 0...n-1.
+// Uses all possible functions in the array proxy interface.
+template <typename T>
+struct Func {
+  template <typename TOut>
+  bool call(TOut& out, const int64_t& n) {
+    for (int i = 0; i < n; i++) {
+      switch (i % 4) {
+        case 0:
+          out.add_item() = i;
+          break;
+        case 1:
+          out.push_back(i);
+          break;
+        case 2:
+          out.add_null();
+          break;
+        case 3:
+          out.resize(out.size() + 1);
+          out[out.size() - 1] = i;
```

Comment:
This is equivalent to calling push_back, right? Any particular reason we don't support reserve?

Reply:
push_back will internally do  call a reserve to ensureSize()
this one does not. 

I do not see how adding a reserve adds value here, 
since if we use push_back we have to ensure size anyway

Reply:
I was thinking that it is standard practice to reserve the vector, then push_back items to it. It would be nice to allow for the same pattern to be used here.

Reply:
I can make the reserve call public.

Reply:
1) resize() invokes the default constructor or value initializes for std containers
2) it's harder to write out of bounds with `reserve() + push_back()`

so I personally prefer `reserve() + push_back/emplace_back` except in rare cases.

Reply:
for arrayProxy: reserve() + push_back/emplace_back is more expensive, 
since when we do push_back, we need to assert the capacity.

it does reflect on numbers. 
reisize for arrayProxy does not invoke the default constructor.

the other way would be to  add push_back_unchecked_cap + reserve()